mozjs-sys/src/jsimpls.rs

```diff
@@ -404,12 +404,12 @@ impl JSNativeWrapper {
 }
 
 impl RootedBase {
-    unsafe fn add_to_root_stack(&mut self, cx: *mut JSContext, kind: JS::RootKind) {
+    unsafe fn add_to_root_stack(this: *mut Self, cx: *mut JSContext, kind: JS::RootKind) {
         let stack = Self::get_root_stack(cx, kind);
-        self.stack = stack;
-        self.prev = *stack;
+        (*this).stack = stack;
+        (*this).prev = *stack;
 
-        *stack = self as *mut _ as usize as _;
+        *stack = this as usize as _;
```

Comment:
Aside: This cast is a roundtrip. I assume the intent is to expose the provenance.

I don't believe that *should* be necessary, but given that there are currently a lot of uses of `deref` and `deref_mut` on this type, creating `&T` and `&mut T` types across GC barriers, I'm not willing to change it. It looks like the sort of thing someone did to stop the compiler from performing some optimizations.

Reply:
I am pretty sure there was a note in one of these cast hoops, but I cannot seem to find it.

Reply:
Note that stack is a pointer to a u64 (possibly u32 on 32 bit targets?) ([source](https://doc.servo.org/mozjs/jsapi/struct.RootingContext.html#structfield.stackRoots_)). It may be that the usize intermediary cast was required by the compiler at some point, but I just checked this snippet on play.rust-lang.org and it doesn't seem like it's needed:
```
    let a = 5;
    let b = &a as *const _ as u64;
    let c = &a as *const _ as usize as u64;
```

Reply:
That usize -> _ cast was introduced in https://github.com/servo/mozjs/commit/d40cfd5ac52aee5bee5b3a19aee387e81837c9fb#diff-e27e06105da7e8ebf851404e43b36501f85aaf53097cdccf96b9f86c8a7cf1daR392 and I'm pretty sure it was needed since stackRoots_ is a u64 value, so it was doing usize -> u64.